test/src/main/java/org/apache/accumulo/test/compaction/ExternalCompactionProgressIT.java

```diff
@@ -0,0 +1,166 @@
+ */
+public class ExternalCompactionProgressIT extends AccumuloClusterHarness {
+  private static final Logger log = LoggerFactory.getLogger(ExternalCompactionProgressIT.class);
+  private static final int ROWS = 10_000;
+
+  enum EC_PROGRESS {
+    STARTED, QUARTER, HALF, THREE_QUARTERS
+  }
+
+  Map<String,RunningCompactorInfo> runningMap = new HashMap<>();
+  List<EC_PROGRESS> progressList = new ArrayList<>();
+
+  private final AtomicBoolean compactionFinished = new AtomicBoolean(false);
+
+  @Override
+  public void configureMiniCluster(MiniAccumuloConfigImpl cfg, Configuration coreSite) {
+    ExternalCompactionTestUtils.configureMiniCluster(cfg, coreSite);
+  }
+
+  @Test
+  public void testProgress() throws Exception {
+    MiniAccumuloClusterImpl.ProcessInfo c1 = null, coord = null;
+    String table1 = this.getUniqueNames(1)[0];
+    try (AccumuloClient client =
+        Accumulo.newClient().from(getCluster().getClientProperties()).build()) {
+      ExternalCompactionTestUtils.createTable(client, table1, "cs1");
+      ExternalCompactionTestUtils.writeData(client, table1, ROWS);
+      c1 = ((MiniAccumuloClusterImpl) getCluster()).exec(Compactor.class, "-q", "DCQ1");
+      coord = ExternalCompactionTestUtils.startCoordinator(((MiniAccumuloClusterImpl) getCluster()),
+          CompactionCoordinator.class, getCluster().getServerContext());
+
+      Thread checkerThread = startChecker();
+      checkerThread.start();
+
+      IteratorSetting setting = new IteratorSetting(50, "Slow", SlowIterator.class);
+      SlowIterator.setSleepTime(setting, 1);
+      client.tableOperations().attachIterator(table1, setting,
+          EnumSet.of(IteratorUtil.IteratorScope.majc));
+      log.info("Compacting table");
+      ExternalCompactionTestUtils.compact(client, table1, 2, "DCQ1", true);
+      ExternalCompactionTestUtils.verify(client, table1, 2, ROWS);
+
+      log.info("Done Compacting table");
+      compactionFinished.set(true);
+      checkerThread.join();
+
+      verifyProgress();
+    } finally {
+      ExternalCompactionTestUtils.stopProcesses(c1, coord);
+    }
+  }
+
+  public Thread startChecker() {
+    return Threads.createThread("RC checker", () -> {
+      try {
+        while (!compactionFinished.get()) {
+          checkRunning();
+          sleepUninterruptibly(1000, TimeUnit.MILLISECONDS);
+        }
+      } catch (TException e) {
+        log.warn("{}", e.getMessage(), e);
+      }
+    });
+  }
+
+  /**
+   * Check running compaction progress.
+   */
+  private void checkRunning() throws TException {
+    var ecList = ExternalCompactionTestUtils.getRunningCompactions(getCluster().getServerContext());
+    var ecMap = ecList.getCompactions();
+    if (ecMap != null) {
+      ecMap.forEach((ecid, ec) -> {
+        // returns null if it's a new mapping
+        var rci = new RunningCompactorInfo(System.currentTimeMillis(), ecid, ec);
+        var previous = runningMap.put(ecid, rci);
```

Comment:
This could potentially return null if the previously assigned value was null, meaning its not a new mapping. Not sure if that would ever happen though.

Reply:
It shouldn't ever store a null value for the `RunningCompactorInfo` object. I replaced var with the actual type and renamed the variable to make it more clear.